azure-keyvault/azure/keyvault/custom/http_message_security.py

```diff
@@ -0,0 +1,169 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+
+import json
+import time
+import uuid
+import os
+from .internal import _a128cbc_hs256_encrypt, _a128cbc_hs256_decrypt, JwsHeader, JwsObject, \
+    JweHeader, JweObject, _str_to_b64, _b64_to_str, _bstr_to_b64, _b64_to_bstr
+from .rsa_key import RsaKey
+
+
+def generate_pop_key():
+    return RsaKey.generate()
+
+
+class HttpMessageSecurity(object):
+    def __init__(self, client_security_token=None,
+                 client_signature_key=None,
+                 client_encryption_key=None,
+                 server_signature_key=None,
+                 server_encryption_key=None):
+        self.client_security_token = client_security_token
+        self.client_signature_key = client_signature_key
+        self.client_encryption_key = client_encryption_key
+        self.server_signature_key = server_signature_key
+        self.server_encryption_key = server_encryption_key
+
+    def protect_request(self, request):
+        # Setup the auth header on the request
+        # Due to limitations in the service we hard code the auth scheme to 'Bearer' as the service will fail with any
+        # other scheme or a different casing such as 'bearer', once this is fixed the following line should be replaced:
+        # request.headers['Authorization'] = '{} {}'.format(auth[0], auth[1])
+        request.headers['Authorization'] = '{} {}'.format('Bearer', self.client_security_token)
+
+        # if the current message security doesn't support message protection, or the body is empty
+        # skip protection and return the original request
+        if not self.supports_protection() or len(request.body) == 0:
+            return request
+
+        plain_text = request.body
+
+        # if the client encryption key is specified add it to the body of the request
+        if self.client_encryption_key:
+            # note that this assumes that the body is already json and not simple string content
+            # this is true for all requests which currently support message encryption, but might
+            # need to be revisited when the types of
+            body_dict = json.loads(plain_text)
+            body_dict['rek'] = self.client_encryption_key.to_jwk().serialize()
+            plain_text = json.dumps(body_dict).encode()
```

Comment:
Empty `encode()` only if you're 100% sure the content is pure ASCII. Otherwise, you should probably be explicit and use UTF8